File: Python3/201 - 400/345 Reverse Vowels of a String.py
# 68ms, 96.88%
class Solution:
    def reverseVowels(self, s):
        s = list(s)
        beg, end = 0, len(s) - 1
        vowels = 'aeiouAEIOU'
        for _ in s:
            if beg >= end: break
            if s[beg] not in vowels:
                beg += 1
                continue
            if s[end] not in vowels:
                end -= 1
                continue
            s[beg], s[end] = s[end], s[beg]
            beg += 1
            end -= 1
        return ''.join(s)
    

File: 201 - 400/345. Reverse Vowels of a String.cc
// 12ms, 80.45%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string reverseVowels(string s) {
        int beg = -1, end = s.size();
        string vowels = "aeiouAEIOU";
        while (beg < end) {
            beg = s.find_first_of(vowels, beg + 1);
            end = s.find_last_of(vowels, end - 1);
            if (beg < end) swap(s[beg], s[end]);
        }
        return s;
    }
};
